File: 814BinaryTreePruning/Solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Returns the root of a pruned tree, or None if the whole tree is made up of 0's or
# None 
def pruneTree(root):
    if not root:
        return None
    
    root.left = pruneTree(root.left)
    root.right = pruneTree(root.right)
    
    # Whether this node can/must be removed
    prunethis = (root.val==0) and (not root.left) and (not root.right)
    return None if prunethis else root

File: 814BinaryTreePruning/Solution.java
import java.util.*;


public class Solution {

    class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode() {}
      TreeNode(int val) { this.val = val; }
      TreeNode(int val, TreeNode left, TreeNode right) {
          this.val = val;
          this.left = left;
          this.right = right;
        }
    }

    boolean shouldremove(TreeNode node) {
        if(node == null) return true;

        boolean removeleft = shouldremove(node.left);
        boolean removeright = shouldremove(node.right);

        // If this node does not belong
        if(removeleft && removeright && node.val == 0) return true;

        // This node belongs, but its subtrees may need to be pruned 
        if(removeleft) node.left = null;
        if(removeright) node.right = null;

        return false;
    }

    TreeNode prunetree(TreeNode root) {
        // Either the whole tree consists of 0's or is null - we need to remove everything+, in that case, just return null
        // Or the root is part of a tree that contains a 1 and should not be removed
        if(shouldremove(root)) return null;

        return root;
    }


	public static void main(String[] argv) {

 	} 	
}